unmarshaller.go

```diff
@@ -1,117 +1,120 @@
-package gocsv
+package commando
 
 import (
-	"encoding/csv"
 	"fmt"
+	"io"
 	"reflect"
 )
 
 // Unmarshaller is a CSV to struct unmarshaller.
 type Unmarshaller struct {
-	reader                 *csv.Reader
-	Headers                []string
-	fieldInfoMap           []*fieldInfo
-	MismatchedHeaders      []string
-	MismatchedStructFields []string
-	outType                reflect.Type
+	config *validConfig
+	line   int
+	reader Reader
 }
 
-// NewUnmarshaller creates an unmarshaller from a csv.Reader and a struct.
-func NewUnmarshaller(reader *csv.Reader, out interface{}) (*Unmarshaller, error) {
+// NewUnmarshaller is a convenience function which allocates and
+// returns a new Unmarshaller.
+func NewUnmarshaller(holder interface{}, reader Reader) (*Unmarshaller, error) {
+	return (&Config{Holder: holder}).NewUnmarshaller(reader)
+}
+
+// NewUnmarshaller creates an unmarshaller from a Reader and a struct.
+func (c *Config) NewUnmarshaller(reader Reader) (*Unmarshaller, error) {
 	headers, err := reader.Read()
 	if err != nil {
 		return nil, err
 	}
-	headers = normalizeHeaders(headers)
 
-	um := &Unmarshaller{reader: reader, outType: reflect.TypeOf(out)}
-	err = validate(um, out, headers)
+	vc, err := c.validate(headers)
 	if err != nil {
 		return nil, err
 	}
+
+	um := &Unmarshaller{
+		reader: reader,
+		config: vc,
+		line:   1,
+	}
+
 	return um, nil
 }
 
-// Read returns an interface{} whose runtime type is the same as the struct that
-// was used to create the Unmarshaller.
+// Read returns an interface{} whose runtime type is the same as the
+// struct that was used to create the Unmarshaller.
 func (um *Unmarshaller) Read() (interface{}, error) {
-	row, err := um.reader.Read()
+	value, _, err := um.ReadUnmatched()
+	return value, err
+}
+
+// ReadAllCallback calls cb for every record Read() from um.
+func ReadAllCallback(um *Unmarshaller, onSuccess func(interface{}) error, onError func(error) error) error {
```

Comment:
I'm not entirely sure what this gives us. I think we could pretty trivially implement this logic at callsites by successively calling `Read` directly until we hit `errors.Is(io.EOF)`, with no loss of flexibility.

This is pretty close to the API [`csv.Reader`](https://pkg.go.dev/encoding/csv#Reader) itself provides. Having a `ReadAll` helper might be convenient, but I'd lean towards returning a slice from `ReadAll` (potentially at the cost of some flexibility), and expect to do the looping ourselves when we need more control.

Reply:
> I'm not entirely sure what this gives us. I think we could pretty trivially implement this logic at callsites by successively calling `Read` directly until we hit `errors.Is(io.EOF)`, with no loss of flexibility.

I'm not sure I agree, but I'll see how things shake out after another pass or two at this.  I wrote this helper function after implementing the loop-and-handle pattern a couple times, and finding it unsatisfactory.

The callback specifically gives us:

1. No possibility of missing that you have to check for `io.EOF` and writing an accidental infinite loop.
2. `onError` allows reuse of error-handling strategies, instead of duplicating them inline everywhere we unmarshal.

Callers are still able to `Read()` in a loop where that makes sense.

> This is pretty close to the API [`csv.Reader`](https://pkg.go.dev/encoding/csv#Reader) itself provides. Having a `ReadAll` helper might be convenient, but I'd lean towards returning a slice from `ReadAll` (potentially at the cost of some flexibility), and expect to do the looping ourselves when we need more control.

It's intentionally close to (and relies on) `csv.Reader`'s API, except it pops out structs instead of `[][]string` which you have to manually process.

I explicitly chose not to implement a `ReadAll()` style function because these have been problematic for us in the past.  For one, we have files large enough that we can't hold the whole thing in memory, and have worked around that by breaking them up ahead of time — but incremental parsing is a better solution, because we don't *need* the whole thing in memory at once, ever.

Additionally, many of the places we parse unmarshal into `[]ConcreteType`, but return `[]InterfaceType`.  So even if we allocate an appropriate slice and return it as `[]interface`, callers would have to traverse it a second time to build the interface slice.

In short, while it seems like a simple and obvious thing to have, I think it has hidden drawbacks, so biased towards the eat-your-veggies approach of forcing incremental parsing, but making it a bit easier.

*But*, I also felt that having a solid base of 1:1 csv-record:struct would let us build that type of thing on top, if we need.

Reply:
Yep - this all makes sense. Agreed that `ReadAll` is a bad approach for big files. A few more thoughts here, around a couple (potentially foolish) potential refinements:

> Additionally, many of the places we parse unmarshal into []ConcreteType, but return []InterfaceType.

I think we might be able to get around this if we pass in a `holder` slice, which could presumably be a slice of any type the concrete type in the config fulfills. There's a chance this would make the reflection bits way worse - I'm definitely not an expert on the details here.

> ...incremental parsing is a better solution, because we don't need the whole thing in memory at once, ever.

If we wanted to be even more clever, I could imagine passing in channel, instead of a slice, as the `holder`, then sending records down the channel as they're parsed.

Putting it all together, we could have something like:

```go
func (um *Unmarshaller) ReadAll(holder interface{}, onError ...func(error) error) error
```

where `holder` could be a `[]ConcreteType`, a `[]InterfaceType`, a `chan<- ConcreteType`, or a `chan<- InterfaceType`. Which would be a really concise API, at the cost of added internal complexity.

My gut feeling is that the channel bit is the bad kind of clever, unless it happens to be easier than I expect. The same is probably true about the `[]InterfaceType`.

Reply:
Also, this is all to unmarshal a CSV, so we've likely already hit the point of diminishing returns in terms of API polish 🙃.

Reply:
Yeah, I agree that would be nice, but I also agree that it's complicated and unlikely to be used.  Gocsv has a bunch of this kind of stuff, which we have never needed.  But we do need a parser that can handle large & slightly broken files without blowing up.